### DISCUSSION TITLE
Can you write migrations in Rust?

### DISCUSSION THREAD

**Discussion by gepbird** [diesel-rs/diesel#3706] [Q&A] [answered]
I'm new to Rust/Diesel and I'm coming from PHP/Laravel where you make migrations in PHP rather than SQL, for example:
```php
    public function up(): void
    {
        Schema::create('flights', function (Blueprint $table) {
            $table->id();
            $table->string('name');
            $table->string('airline');
            $table->timestamps();
        });
    }

    public function down(): void
    {
        Schema::drop('flights');
    }
```

Is it possible with Diesel? I only found SQL migrations in the examples provided. I found this piece of code that creates a migration but I'm unsure how to run it:
```rust
// migrations/20230719120000_create_table_name.rs
use diesel::migration::Migration;
use diesel::sql_types::*;
use diesel::*;

table! {
    users (id) {
        id -> Integer,
        name -> Text,
    }
}

pub struct CreateTable;

impl Migration for CreateTable {
    fn up(&self) {
        users::table
            .create(&self.connection())
            .expect("Error creating table_name table");
    }

    fn down(&self) {
        users::table
            .drop(&self.connection())
            .expect("Error dropping table_name table");
    }
}
```
Thanks for your help in advance.

**Comment by weiznich** [comment, marked as answer]
Technically speaking that's already possible because a [`Migration`](https://docs.diesel.rs/2.1.x/diesel/migration/trait.Migration.html) trait with the relevant methods already exists. Practically speaking there is no real documentation how to do that and there are now helper methods to simply create/modify/… a table. I would be open for contributions here, just be sure to first write down some ideas before starting to implement stuff. (There is also some existing discussion about that here: https://github.com/diesel-rs/diesel/pull/3373)

Now there is an different way to approach this: Diesel-CLI allows you to generate migrations based on a changes in your `schema.rs` file compared with your running database schema. That behind the `diesel migration generate --diff-schema` flag. It will generate a SQL migration that can modified before applying.
